### PR TITLE
[compiler] InferReferenceEffects outputs a disjoint set of aliases

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30976
* #30975
* __->__ #30974
* #30973
* #30972

Test Plan:
For the purposes of a pass added later in this stack, the InferReferenceEffects now outputs a DisjointSet of identifier ids representing aliases. We add this to InferReferenceEffects because this pass already does its own alias analysis, so making it available to later passes is convenient.

This entails implementing copy() and equals() methods on DIsjointSets so that the InferenceState copy and merge methods can handle alias sets.

### CODE REVIEW

**Review by mofeiZ — COMMENTED** [review]
Nit: It might be helpful to be explicit about what this alias map represents with a comment / docblock . iiuc InferReferenceEffects aliases are conservative (no false positives, maybe false negatives). InferReactiveScopeVariables on the other hand uses maybe-aliases (e.g. pessimistic on function calls)

Also had a qq: reading through InferReferenceEffects, I notice that we [currently alias](https://github.com/facebook/react/blob/gh/mvitousek/34/head/compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts#L1811-L1820) each destructuring target of a `Destructure` instruction to source place.

It seems that all other `alias` callsites are for known / precise aliases (LoadLocal and StoreLocals, evaluation of a PropertyStore). Similarly, we don't alias PropertyLoads. What's the difference between Destructure and PropertyLoad here?

File: compiler/packages/babel-plugin-react-compiler/src/Utils/DisjointSet.ts [hunk 1]
```diff
@@ -124,6 +124,42 @@ export default class DisjointSet<T> {
...
+      }
+    }
+    for (const item of other.#entries.keys()) {
+      const thisRoot = this.find(item);
+      if (thisRoot === null) {
+        return false;
+      }
+    }
+    return true;
+  }
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Utils/DisjointSet.ts:161]: can you add some tests for this?

imagine two Sets:
1. `this` has groups `{a}`, `{b}`
2. `other` has one group, `{a, b}`

we'd iterate this:
* `a`. otherRoot = a or b, thisRoot = a, rootMap doesn't have a so we continue
* `b`. otherRoot = a or b, thisRoot = b, rootMap doesn't have b so we continue

now we iterate other:
* `a`, thisRoot = a which is non-null so continue
* `b`, thisRoot = b which is non-null so continue

So i think this example ends up incorrectly returning true (equals) when they're not equal.

**Review by josephsavona — CHANGES_REQUESTED** [review]
Seems fine to return the alias sets (though i agree w @mofeiZ about adding some comments for what it represents). But I think the equals() method might have a bug, can you add some extra tests (jest tests, not normal snap tests) to double-check?

**Comment by mvitousek** [comment]
Fixed, thanks for the catch! Now build up the rootMap before comparing the keys. Added tests as well.

**Pull request closed without merging**